test/opt/test_def_use.cpp

```diff
@@ -0,0 +1,599 @@
+  SpvTools tools(SPV_ENV_UNIVERSAL_1_1);
+
+  std::vector<uint32_t> binary;
+  // We need this to generate the necessary header in the binary.
+  tools.Assemble("", &binary);
+  inst->ToBinary(&binary, /* skip_nop = */ false);
+
+  std::string text;
+  // We'll need to check the underlying id numbers.
+  // So turn off friendly names for ids.
+  tools.Disassemble(binary, &text, SPV_BINARY_TO_TEXT_OPTION_NO_HEADER);
+  while (!text.empty() && text.back() == '\n') text.pop_back();
+  return text;
+}
+
+// A struct for holding expected id defs and uses.
+struct InstDefUse {
+  using IdInstPair = std::pair<uint32_t, const char*>;
+  using IdInstsPair = std::pair<uint32_t, std::vector<const char*>>;
+
+  // Ids and their corresponding def instructions.
+  std::vector<IdInstPair> defs;
+  // Ids and their corresponding use instructions.
+  std::vector<IdInstsPair> uses;
+};
+
+// Checks that the |actual_defs| and |actual_uses| are in accord with
+// |expected_defs_uses|.
+void CheckDefUse(const InstDefUse& expected_defs_uses,
+                 const DefUseManager::IdToDefMap& actual_defs,
+                 const DefUseManager::IdToUseMap& actual_uses) {
+  // Check defs.
+  EXPECT_EQ(expected_defs_uses.defs.size(), actual_defs.size());
+  for (uint32_t i = 0; i < expected_defs_uses.defs.size(); ++i) {
+    const auto id = expected_defs_uses.defs[i].first;
+    const auto expected_def = expected_defs_uses.defs[i].second;
+    ASSERT_EQ(1u, actual_defs.count(id));
+    EXPECT_EQ(expected_def, DisassembleInst(actual_defs.at(id)));
+  }
+
+  // Check uses.
+  EXPECT_EQ(expected_defs_uses.uses.size(), actual_uses.size());
+  for (uint32_t i = 0; i < expected_defs_uses.uses.size(); ++i) {
+    const auto id = expected_defs_uses.uses[i].first;
+    const auto& expected_uses = expected_defs_uses.uses[i].second;
+
+    ASSERT_EQ(1u, actual_uses.count(id));
+    const auto& uses = actual_uses.at(id);
+
+    EXPECT_EQ(expected_uses.size(), uses.size());
+    auto it = uses.cbegin();
+    for (const auto expected_use : expected_uses) {
+      EXPECT_EQ(expected_use, DisassembleInst(it->first));
+      ++it;
+    }
+  }
+}
+
+struct ParseDefUseCase {
+  const char* text;
+  InstDefUse du;
+};
+
+using ParseDefUseTest = ::testing::TestWithParam<ParseDefUseCase>;
+
+TEST_P(ParseDefUseTest, Case) {
+  const auto& tc = GetParam();
+
+  // Build module.
+  const std::vector<const char*> text = {tc.text};
+  std::unique_ptr<ir::Module> module =
+      SpvTools(SPV_ENV_UNIVERSAL_1_1).BuildModule(JoinAllInsts(text));
+  ASSERT_NE(nullptr, module);
+
+  // Analyze def and use.
+  opt::analysis::DefUseManager manager;
+  manager.AnalyzeDefUse(module.get());
+
+  CheckDefUse(tc.du, manager.id_to_defs(), manager.id_to_uses());
+}
+
+// clang-format off
+INSTANTIATE_TEST_CASE_P(
+    TestCase, ParseDefUseTest,
+    ::testing::ValuesIn(std::vector<ParseDefUseCase>{
+        {"", {{}, {}}},                              // no instruction
```

Comment:
Please add a test case where an OpPhi is both the definintion and use of a value.

Note that at some point LLVM decided that mention of a value on the right-hand-side of a phi is *not* considered a use.  But our code does consider it a use.
(And we might want to make that clear in the comments for the DefUseManager)